_resources/port1.0/group/java-1.0.tcl

```diff
@@ -154,4 +134,103 @@ proc java_set_env {} {
     destroot.env-append    JAVA_HOME=${java_home}
     java.home ${java_home}
 }
+
 port::register_callback java_set_env
+
+proc get_jvm_bigsur { version_requested } {
+    # Normalize version numbers 1.x -> x
+    set version_requested [regsub {^1\.} $version_requested ""]
+    # Sort the JVMs we found on the system descending
+    set versions_found [sort_dict [find_jvm_versions]]
+    ui_debug "java-portgroup: Detected JVMs: $versions_found"
+    # Match the systems JVMs with the one requested by MacPorts
+    # Higher JVM Versions win
+    return [match_jvm_version $version_requested $versions_found]
+}
+
+# Find JVM versions by using /usr/libexec/java_home -V (not -v!)
+# Returns a dict with unique major versions as key and the corresponding
+# java_home as value:
+#  dict[major_version]=java_home
+proc find_jvm_versions {} {
+    if {[catch {exec /usr/libexec/java_home -V} result options]} {
+        # Extract JVM versions and corresponding JAVA_HOMEs
+        set vm_versions [regexp -all -inline -- { +(\d+(?:\.\d+)+)[^/]+(\/[^\0\n]+)} $result]
+        # %3=0 -> Regex match, ignored.
+        # %3=1 -> Version
+        # %3=2 -> JAVA_HOME.
+        for {set idx 0} {$idx < [llength $vm_versions]} {incr idx 3} {
+            set vers [lindex $vm_versions $idx+1]
+            # Normalize version 1.x -> x
+            set vers [regsub {^1\.} $vers ""]
+            # Extract major version
+            set vers [regsub {(\.\d+)+} $vers ""]
+            set path [lindex $vm_versions $idx+2]
+            dict append version_path_dict $vers $path
+        }
+    } else {
+        set details [dict get $options -errorcode]
+        ui_debug "Error executing /usr/libexec/java_home -V"
+    }
+    set version_path_dict
+}
+
+# Match a normalized major version string like 7, 8*, 9+
+proc match_jvm_version { version_requested versions_found } {
+    if {[string first "+" $version_requested] != -1} {
+        set version_requested [regsub {\+} $version_requested ""]
+        match_less_equal $version_requested $versions_found
+    } else {
+        set version_requested [regsub {\*} $version_requested ""]
+        match_equal $version_requested $versions_found
+    }
+}
+
+# Returns the value of the first dictionary entry
+# whose key is equal to search_key.
+#
+# @param search_key The key that is to be found in the dictionary
+# @param target_dict The dictionary which should be searched for search_key
+# @return The value of the first entry that matched.
+# Returns an error, if no such entry is found.
+proc match_equal { search_key target_dict } {
```

Comment:
Would you mind still changing these generic names (e.g. `s/match_equal/match_jvm_equal/g`) to avoid polluting the global namespace, or placing all the utility functions in an actual Tcl namespace? It's handy that they're generically useful and not specific to the java portgroup, but until such a time as they're pulled into a common library location for general use, having them here just risks unexpectedly introducing/stomping on names in the global namespace.

Reply:
I wrapped all utility functions into a 'java' namespace.